### ISSUE TITLE
Cipher State Serialization

### ISSUE THREAD

**Issue by slontis** [openssl/openssl#14223]
Functionality to load and save the internal state of a cipher would be useful for systems using multi processors with fail-over capabilities. This would allow the user to save a blob of data any time after the init has completed.

A possible implementation would need to do the following:

    High level API's to serialize and deserialize an EVP_CIPHER_CTX.
    Each provider cipher implementation would need to have methods to serialize/deserialize their internal data.
    Have some blob format.

Some issues related to the blob format (if the data was not ephemeral):

    Handling backwards compatibility would complicate the serialization (since any small change to a digest would cause compatibility problems). Any additional new fields in a digest struct could potentially behave differently and have no way of working with older code.
    A different provider could possibly serialize a cipher differently.
    A different implementation (e.g uses different assembler) could serialize differently.

**Comment by levitte** [comment]
Is this idea something similar to what `EVP_CIPHER_param_to_asn1()` and `EVP_CIPHER_asn1_to_param()` do, but a bit more involved?

**Comment by kaduk** [comment]
If I understand correctly, this is already possible (*) by using the EVP_CIPHER_CTX_get_updated_iv() API and the raw key; EVP_CipherInit() using the "updated" IV would then be able to resume processing where the old one left off.

(*) I think it requires that the input already processed ended on a block boundary.

**Comment by paulidale** [comment]
> Is this idea something similar to what `EVP_CIPHER_param_to_asn1()` and `EVP_CIPHER_asn1_to_param()` do, but a bit more involved?

The idea is to stash the state of a cipher context after some updates and to be able to recover it later elsewhere.
I'm pretty sure `EVP_CIPHER_param_to_asn1()` doesn't do this.

**Comment by beldmit** [comment]
> If I understand correctly, this is already possible (*) by using the EVP_CIPHER_CTX_get_updated_iv() API and the raw key; EVP_CipherInit() using the "updated" IV would then be able to resume processing where the old one left off.
> 
> (*) I think it requires that the input already processed ended on a block boundary.

It may be much more complicated than just iv and key. At least, the amount of data processed may matter, and many other things. This option is highly implementation-dependent, so I'd think it's a post-3.0 task

**Comment by paulidale** [comment]
Absolutely post 3.0, if ever.

**Comment by levitte** [comment]
I'm trying to understand what should happen with that blob.  Is it something you'd expect to store away somewhere, to re-use the next day with a new EVP_CIPHER_CTX?  Or is this basically a new way to do EVP_CIPHER_CTX_copy()?

**Comment by paulidale** [comment]
The former is the closest as I understand it.

**Comment by slontis** [comment]
The use case is a blob of data is saved to shared memory so that another process could take the blob and create a new ctx.

**Comment by mdpung** [comment]
Please correct me in any misunderstandings, but in terms of block encryption algorithms that use counters like AES-GCM and ChaCha20, I understand there are some workarounds to this with keeping track of the number of blocks written to the cipher. Yet, I can imagine it would be more performant and efficient for the library to provide the functionality to store this type of state information and be able to load it back into the cipher without the overhead of manually cycling through these redundant encryption operations in order to match the value of the counter.

**Comment by davidben** [comment]
No, that's not quite how that works. There is no need to manually cycle through redundant operations.

For an AEAD like AES-GCM or ChaCha20-Poly1305, there are two things that look like counters. The first is that, _within_ a single encryption operation (i.e. for a single IV, or nonce), individual blocks are encrypted with incrementing counters. However, that is all within a _single_ encryption operation, so it would not make sense to save and restore that value.

The second is that _distinct_ encryptions use distinct IVs. In many applications, this IV is constructed with a counter, such as the sequence number of a TLS record. However, this is a _parameter_ to the encryption operation, not state maintained within the encryption operation. As such, there is nothing to serialize here.

Now, although IVs are external parameters, OpenSSL does have some odd APIs like `EVP_CTRL_GCM_IV_GEN` which internally managing a counter for generating an IV. However, even there, there is no need to manually cycle things because those APIs allow you to start the counter at a particular point.

This is generally true. Encryption algorithms are generally designed so that, where it makes sense to seek ahead, doing so is part of their public interface, without the need of some extra state serialization operation. RC4 is the only instance I can think of that breaks this, but RC4 is broken and no one should be making design decisions based on it anymore. :-)

**Comment by mdpung** [comment]
> No, that's not quite how that works. There is no need to manually cycle through redundant operations.
> 
> For an AEAD like AES-GCM or ChaCha20-Poly1305, there are two things that look like counters. The first is that, _within_ a single encryption operation (i.e. for a single IV, or nonce), individual blocks are encrypted with incrementing counters. However, that is all within a _single_ encryption operation, so it would not make sense to save and restore that value.
> 
> The second is that _distinct_ encryptions use distinct IVs. In many (but not all!) applications, this IV is constructed with a counter, such as the sequence number of a TLS record. However, this is a _parameter_ to the encryption operation, not state maintained within the encryption operation. As such, there is nothing to serialize here.
> 
> Now, although IVs are external parameters, OpenSSL does have some odd APIs like `EVP_CTRL_GCM_IV_GEN` which internally managing a counter for generating an IV. However, even there, there is no need to manually cycle things because those APIs allow you to start the counter at a particular point. They're also equivalent to just using the more standard AEAD interface, so you can also just not use them.
> 
> This is generally true. Encryption algorithms are generally designed so that, where it makes sense to seek ahead, doing so is part of their public interface, without the need of some extra state serialization operation. RC4 is the only instance I can think of that breaks this, but RC4 is broken and no one should be making design decisions based on it anymore. :-)

Ah, I think I see where the confusion is. I've seen GCM referred to behaving like a Stream Cipher, so in this context I've been trying to use it as one. I think this is where my view point is flawed. You mention GCM encryption occurs `all within a single encryption operation`, and I'm trying to use it as an actual Stream Cipher where I want the block cipher to continue. Sounds like I'm trying to make a dog bark like a cat. I'll look into actual stream ciphers for my use case.

Thank you for your response!

**Comment by mdpung** [comment]
It appears there are no NIST approved Stream Cipher algorithms, which makes me want to loop back to looking at block ciphers for my use case.

To clarify such use case, imagine each block is a full payload of 128 bits, and there are an arbitrary number of payloads to be encrypted. Rather than doing a complete encryption operation on each block and seemingly not making use of the counter, I want to maintain the state of the block cipher IV so it can continue to utilize the counter functionality for a unique IV for each payload.

**Comment by davidben** [comment]
AES-GCM is indeed not a stream cipher. If you use it as that, you have a security bug because you're actually just getting a slightly different variant of CTR mode, and not getting the security properties of AES-GCM at all. AES-GCM is an authenticated encryption, so there's a tag that's computed to authenticate things. That kind of thing is fundamentally incompatible with streaming.

It's just that OpenSSL's API for AES-GCM is confusing and dangerous and makes it very easy to mistakenly stream an AES-GCM operation.

**Comment by lin755** [comment]
Can we explain it that way? for aead update mode, data is associated between update stage. but in a tls mode, IV is updated only to maintain uniqueness.

**Comment by lin755** [comment]
update case :                             tls case:
EVP_EncryptInit_ex                     EVP_EncryptInit_ex
EVP_EncryptUpdate                   EVP_CIPHER_CTX_ctrl(EVP_CTRL_GCM_SET_IV_FIXED)
EVP_EncryptUpdate                   EVP_CIPHER_CTX_ctrl(EVP_CTRL_AEAD_TLS1_AAD)
EVP_EncryptFinal_ex                  EVP_EncryptUpdate    //single operation, but update iv when IV_GEN
                                                  EVP_EncryptUpdate    //update iv

I‘m not sure if the tls case is right, is multiple update allowed?

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
